### PR TITLE
update evaluator.py

### PR DESCRIPTION
**Pull request by Laughing-q**
I just noticed I commented the sort code in mask_nms...I tested the sorted results first, and then the unsorted results, finally forgot to cancel the comment...
Also I update the evaluator.py then you can set `plots=True` with `mask_ratio=1,2,4,8...` during training.

### CODE REVIEW

**Comment by AyushExel** [comment]
@Laughing-q Ok thanks.. I think the attempt to fix memory leak didn't work. But I'll test again without plots to make sure plotting is the issue..then I'll move on to test plotting with mask-ratio 4

**Comment by Laughing-q** [comment]
@AyushExel I noticed some unusual CUDA memory growth in my experiments, I think this might also be related with the number of `gt_masks`. Cause each `gt_mask` would be a 640x640 img when we set `mask_ratio=1` and the number of `gt_masks` counld be a lot and variable with mosaic augmentation and yolov5's label assign.

**Comment by AyushExel** [comment]
@Laughing-q hmm.. The number of ground truth mask is affected by the mask ratio? Or just the size? 

Yeah I agree the growth in CUDA memory( and even system memory ) is quite strange

**Comment by Laughing-q** [comment]
> @Laughing-q hmm.. The number of ground truth mask is affected by the mask ratio? Or just the size? 
> 
> Yeah I agree the growth in CUDA memory( and even system memory ) is quite strange

just the size.

**Comment by AyushExel** [comment]
@Laughing-q Do you know what mask-ratio other tools like mmdet, paddledet or detectron2 recommend training at? Do they also recommend training with mask-ratio 4 or is it 1?

**Comment by Laughing-q** [comment]
@AyushExel I didn't see `mask_ratio` in other tools, actually I created this argument. The original intention of designing `mask_ratio` is that yolov5 will assign more targets than other detection frameworks, and the corresponding gt_masks will occupy a lot of cuda memory during training, so I add this argument for saving cuda memory.

Also I didn't see plotting during training in other tools(I'm not so sure), so maybe they don't have our plotting issue.

**Comment by AyushExel** [comment]
@Laughing-q plotting is not the major issue.. I think memory leak is happening without plotting..Also, we're plotting mosaics which is slower than plotting a batch of images( add a ceil of 8-10 images per epoch max) .. right now its too much.

I'm more interested in what you said about targets:

> yolov5 will assign more targets than other detection frameworks

Why is this so? Can we reduce the targets itself to eliminate the problem of mask-ratio altogether? Maybe it'll also solve other issues

**Comment by Laughing-q** [comment]
@AyushExel There are two places that yolov5 will assign more targets.
1. mosaic augmentation, concatenate 4 images, which introduce more targets.
2. the yolov5 label assign method not only assigns the center grid, but also assigns the surrounding grids to improve recall.
[https://github.com/ultralytics/yolov5/blob/f8722b4429e80f96be04b36e4efd84ce6583bfa1/utils/loss.py#L217](https://github.com/ultralytics/yolov5/blob/f8722b4429e80f96be04b36e4efd84ce6583bfa1/utils/loss.py#L217)

Other instance segmentation don't use the mosaic augmentation and assign labels through iou(mask-rcnn) or just center grid(solov1/v2).

**Comment by Laughing-q** [comment]
@AyushExel By the way, it seems that plotting will affect cuda memory in my experiments, but I'm not so sure. As I mentioned before, cuda memory changes strangely

**Comment by Laughing-q** [comment]
@AyushExel There might be misunderstanding in the initial PR description. These updates will make `plots=True` compatible with mask_ratio=1 or other values. Previously, plotting and mask ratio=1 were bound.

**Comment by AyushExel** [comment]
@Laughing-q OK got it.. I think that instead of training at mask-ratio 4, we should think of certain optimizations that allow training at ratio 1.. because downsampling will definitely cause lower accuracies..
A few things that I noticed are:
* Training yolov5s takes ~50GB system memory. It takes only 20GB for detection training.. Some increase is explained by mask labels but it should not be this much.. -- Here, we need to look into dataloader
* As you said about targets increasing due to mosaic augmentation, Maybe let's test after disabling that? 
* The loss function is slower by more than a factor of 10.. This can also be optimized.. 

The most important thing here is to optimize system memory..Even after training without plots, it is taking 50GB which I think is too much..

**Comment by AyushExel** [comment]
Screenshot 2022-07-14 at 7 40 01 PM

@Laughing-q I think we should focus on fixing the memory leak in evaluator. See how training for 1 epoch takes only 20GB which same as training detection..but It jumps by double to 45GB during evaluation.. There is no reason for this much increase in consumption of memory.. first, because val set is m much smaller. Second, there is no augmentation in val set, thirdly, the val loader is already loaded before training so it should not take up any more memory.. SO there is no reason for this much memory cost.. it we can fix this, I think plots will not be an issue..
What do you think?

**Comment by AyushExel** [comment]
I'll merge this PR and then let's discuss the above ^^ issue in another thread
